src/app/api/reducer.ts

```diff
@@ -0,0 +1,68 @@
+import {useEffect, useReducer} from "react";
+import {AxiosPromise} from "axios";
+import {Data} from "../components/model/models";
+interface State<P> {
+    data?: P
+    isError: boolean
+    isLoading: boolean
+}
+
+interface Action<P> {
+    type: 'FETCH_INIT' | 'FETCH_ERROR' | 'FETCH_SUCCESS'
+    payload?: P
+}
+
+const fetchReducer = <P>(state: State<P>, action: Action<P>): State<P> => {
+    switch (action.type) {
+        case "FETCH_INIT":
+            return {
+                ...state,
+                isError: false,
+                isLoading: true
+            }
+        case 'FETCH_ERROR':
+            return {
+                ...state,
+                isError: true,
+                isLoading: false
+            }
+        case "FETCH_SUCCESS":
+            return {
+                ...state,
+                data: action.payload,
+                isError: false,
+                isLoading: false
+            }
+        default:
+            throw new Error("Unknown error was encountered while fetching content.")
+    }
+}
+
+export const useFetchAPI = <P>(initData: P, supplier: () => AxiosPromise<Data>, dependencyList:ReadonlyArray<any>=[], transformer?: (data:Data) => P) => {
```

Comment:
No ! Not data ! It's a generic function ! Also, it should not be specific to your application.

Reply:
I added a second generic type (now one is the initial response type and the other is the transformed payload type).

Reply:
Yes good job